### ISSUE TITLE
Quoted `Final` is not "allowed in this context"

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1996]
**Describe the bug**
A quoted `Final` annotation produces a type error, '"Final" not allowed in this context'.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no errors would be produced

**Screenshots or Code**

```python
# pyright: strict

import typing
import typing_extensions

FOO: typing.Final[int] = 1
BAR: "typing.Final[int]" = 2
BAZ: "typing_extensions.Final" = 3

```

**VS Code extension or command-line**
1.1.149

**Comment by erictraut** [comment]
The `Final` mechanism is unfortunate in the way it was designed. It's not a type but rather a type modifier, and unlike other type annotations, it needs to be understood during the "binding" phase of analysis. For that reason, pyright contains a bunch of special-case logic to detect it during the binding phase, but it cannot handle quoted forward references.

There's no easy fix for this, and I don't think it's worth fixing when there's a straightforward workaround: don't use quotes for Final. If you need to use a forward reference, quote the Final type argument only, as in `typing.Final["MyClass"]`.

**Comment by layday** [comment]
I could do that, although I suspect the easier fix would be to not use `Final`, since the purpose of quoting it was to avoid placing a runtime dependency on `typing-extensions`, in a library which needs to support Python 3.6.  Thanks for explaining.
